Tests/SourceKitLSPTests/SwiftCompletionTests.swift

```diff
@@ -547,6 +547,66 @@ final class SwiftCompletionTests: XCTestCase {
                                                 position: Position(line: 7, utf16index: 12),
                                                 context:CompletionContext(triggerKind: .triggerFromIncompleteCompletions)))))
   }
+
+  /// Regression test for https://bugs.swift.org/browse/SR-13561 to make sure the a session
+  /// close waits for its respective open to finish to prevent a session geting stuck open.
+  func testSessionCloseWaitsforOpen() throws {
+    initializeServer(options: SKCompletionOptions(serverSideFiltering: true, maxResults: nil))
+    let url = URL(fileURLWithPath: "/\(#function)/file.swift")
+    openDocument(text: """
+      struct S {
+        func forSomethingCrazy() {}
+        func forSomethingCool() {}
+        func test() {
+          self.forSome
+        }
+        func print() {}
+        func anotherOne() {
+          self.prin
+        }
+      }
+      """, url: url)
+
+    let forSomeComplete = CompletionRequest(
+          textDocument: TextDocumentIdentifier(url),
+        position: Position(line: 4, utf16index: 12), // forS^
+        context:CompletionContext(triggerKind: .invoked))
+    let printComplete = CompletionRequest(
+        textDocument: TextDocumentIdentifier(url),
+      position: Position(line: 8, utf16index: 12), // prin^
+      context:CompletionContext(triggerKind: .invoked))
+    let coordinationGroup = DispatchGroup()
```

Comment:
Using XCTest expectations will behave better during testing.  For one thing, they will provide better behaviour if something goes wrong (timeout indicating which request didn't finish).  Another difference is that if there is anything in the service that needs to use the main thread, XCTest will automatically cycle the main run loop when waiting on an expectation, while blocking would otherwise cause a deadlock in that situation.

Reply:
Done